internal/app/machined/pkg/controllers/secrets/api.go

```diff
@@ -0,0 +1,530 @@
+	"github.com/talos-systems/talos/pkg/resources/secrets"
+	timeresource "github.com/talos-systems/talos/pkg/resources/time"
+	"github.com/talos-systems/talos/pkg/resources/v1alpha1"
+)
+
+// APIController manages secrets.API based on configuration to provide apid certificate.
+type APIController struct{}
+
+// Name implements controller.Controller interface.
+func (ctrl *APIController) Name() string {
+	return "secrets.APIController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *APIController) Inputs() []controller.Input {
+	// initial set of inputs: wait for machine type to be known and network to be partially configured
+	return []controller.Input{
+		{
+			Namespace: network.NamespaceName,
+			Type:      network.StatusType,
+			ID:        pointer.ToString(network.StatusID),
+			Kind:      controller.InputWeak,
+		},
+		{
+			Namespace: config.NamespaceName,
+			Type:      config.MachineTypeType,
+			ID:        pointer.ToString(config.MachineTypeID),
+			Kind:      controller.InputWeak,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *APIController) Outputs() []controller.Output {
+	return []controller.Output{
+		{
+			Type: secrets.APIType,
+			Kind: controller.OutputExclusive,
+		},
+	}
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo
+func (ctrl *APIController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-r.EventCh():
+		}
+
+		machineTypeRes, err := r.Get(ctx, resource.NewMetadata(config.NamespaceName, config.MachineTypeType, config.MachineTypeID, resource.VersionUndefined))
+		if err != nil {
+			if state.IsNotFoundError(err) {
+				continue
+			}
+
+			return fmt.Errorf("error getting machine type: %w", err)
+		}
+
+		machineType := machineTypeRes.(*config.MachineType).MachineType()
+
+		networkResource, err := r.Get(ctx, resource.NewMetadata(network.NamespaceName, network.StatusType, network.StatusID, resource.VersionUndefined))
+		if err != nil {
+			if state.IsNotFoundError(err) {
+				continue
+			}
+
+			return err
+		}
+
+		networkStatus := networkResource.(*network.Status).TypedSpec()
+
+		if !(networkStatus.AddressReady && networkStatus.HostnameReady) {
+			continue
+		}
+
+		// machine type is known and network is ready, we can now proceed to one or another reconcile loop
+		switch machineType {
+		case machine.TypeInit, machine.TypeControlPlane:
+			if err = ctrl.reconcileControlPlane(ctx, r, logger); err != nil {
+				return err
+			}
+		case machine.TypeJoin:
+			if err = ctrl.reconcileJoin(ctx, r, logger); err != nil {
+				return err
+			}
+		case machine.TypeUnknown: // nothig to do
+		}
+
+		if err = ctrl.teardownAll(ctx, r); err != nil {
+			return err
+		}
+	}
+}
+
+//nolint:gocyclo,cyclop
+func (ctrl *APIController) reconcileControlPlane(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
```

Comment:
this and following reconcile methods should be merged, as they have too much in common